Prevent OOB access in foreach_list variants

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112348
* __->__ #112349

By checking that lists sizes are the same before computing forward gradients.

Before the change
```cpp
::std::vector<at::Tensor> _foreach_add_List(c10::DispatchKeySet ks, at::TensorList self, at::TensorList other, const at::Scalar & alpha) {
  auto self_ = unpack(self, "self", 0);
  auto other_ = unpack(other, "other", 1);
  [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self, other );

  std::vector<bool> _any_has_forward_grad_result(self.size());
  for (const auto& i : c10::irange(self.size())) {
    _any_has_forward_grad_result[i] = isFwGradDefined(self[i]) || isFwGradDefined(other[i]);
  }
  ...
```
after the change:
```cpp
::std::vector<at::Tensor> _foreach_add_List(c10::DispatchKeySet ks, at::TensorList self, at::TensorList other, const at::Scalar & alpha) {
    auto self_ = unpack(self, "self", 0);
    auto other_ = unpack(other, "other", 1);
    [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self, other );

    TORCH_CHECK(
        self.size() == other.size(),
          "Tensor lists must have the same number of tensors, got ",
        self.size(),
          " and ",
        other.size());
    std::vector<bool> _any_has_forward_grad_result(self.size());
    for (const auto& i : c10::irange(self.size())) {
      _any_has_forward_grad_result[i] = isFwGradDefined(self[i]) || isFwGradDefined(other[i]);
    }

```
Add regression test

Fixes https://github.com/pytorch/pytorch/issues/112305

malfet committed:
Prevent OOB access in foreach_list variants

By checking that lists sizes are the same before computing forward gradients.

Add regression test

Fixes https://github.com/pytorch/pytorch/issues/112305

[ghstack-poisoned]

malfet committed:
Update on "Prevent OOB access in foreach_list variants"


By checking that lists sizes are the same before computing forward gradients.

Add regression test

Fixes https://github.com/pytorch/pytorch/issues/112305

[ghstack-poisoned]

malfet committed:
Update on "Prevent OOB access in foreach_list variants"


By checking that lists sizes are the same before computing forward gradients.

Add regression test

Fixes https://github.com/pytorch/pytorch/issues/112305

[ghstack-poisoned]

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -712,6 +712,16 @@
+FW_DERIVATIVE_SIZE_CHECK_TEMPLATE = CodeTemplate(
+    """\
+  TORCH_CHECK(
+      self.size() == ${inp_name}.size(),
```
I guess `self` is always a Tensorlist here? is that an ok assumption to make?

albanD reviewed (commented):
Sounds ok, can you share how the updated codegened code looks like?
btw I am wondering if a better fix wouldn't be to move this code that computes `_any_has_forward_grad_result` down to where it is used after the redispatch. That means that we are guaranteed that the op ran and so we should be able to assume that all inputs are valid (otherwise the redispatch would have raised an error). That would avoid use doing duplicate size checks that might become painful.

cc @crcrpar who looked at that code.

malfet commented:
@albanD updated description (and actually, spotted a bug, where current implementation adds one tautological check)

malfet committed:
Update on "Prevent OOB access in foreach_list variants"


By checking that lists sizes are the same before computing forward gradients.

Before the change
```cpp
::std::vector<at::Tensor> _foreach_add_List(c10::DispatchKeySet ks, at::TensorList self, at::TensorList other, const at::Scalar & alpha) {
  auto self_ = unpack(self, "self", 0);
  auto other_ = unpack(other, "other", 1);
  [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self, other );

  std::vector<bool> _any_has_forward_grad_result(self.size());
  for (const auto& i : c10::irange(self.size())) {
    _any_has_forward_grad_result[i] = isFwGradDefined(self[i]) || isFwGradDefined(other[i]);
  }
  ...
```
after the change:
```cpp
::std::vector<at::Tensor> _foreach_add_List(c10::DispatchKeySet ks, at::TensorList self, at::TensorList other, const at::Scalar & alpha) {
    auto self_ = unpack(self, "self", 0);
    auto other_ = unpack(other, "other", 1);
    [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( self, other );
      TORCH_CHECK(
          self.size() == other.size(),
            "Tensor lists must have the same number of tensors, got ",
          self.size(),
            " and ",
          other.size());
    std::vector<bool> _any_has_forward_grad_result(self.size());
    for (const auto& i : c10::irange(self.size())) {
      _any_has_forward_grad_result[i] = isFwGradDefined(self[i]) || isFwGradDefined(other[i]);
    }

```
Add regression test

Fixes https://github.com/pytorch/pytorch/issues/112305

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

crcrpar commented:
thank you for this fix
